packages/mobile/src/web3/contracts.ts

```diff
@@ -1,128 +1,147 @@
+import { sleep } from '@celo/utils/src/async'
+import { call, delay, select } from 'redux-saga/effects'
+import { ErrorMessages } from 'src/app/ErrorMessages'
 import { DEFAULT_FORNO_URL } from 'src/config'
-import { IPC_PATH } from 'src/geth/geth'
-import { waitForGethConnectivity } from 'src/geth/saga'
-import { promisifyGenerator } from 'src/utils/generators'
+import { isProviderConnectionError } from 'src/geth/geth'
+import { waitForGethInitialized } from 'src/geth/saga'
+import { navigateToError } from 'src/navigator/NavigationService'
 import Logger from 'src/utils/Logger'
-import { Actions } from 'src/web3/actions'
-import { contractKitReadySelector, fornoSelector } from 'src/web3/selectors'
+import { getHttpProvider, getIpcProvider } from 'src/web3/providers'
+import { fornoSelector } from 'src/web3/selectors'
 import Web3 from 'web3'
-import { IpcProvider, provider } from 'web3-core'
+import { IpcProvider } from 'web3-core'
 
-const tag = 'web3/contracts'
+const TAG = 'web3/contracts'
+const KIT_INIT_RETRY_DELAY = 2000
 
 let ipcProvider: IpcProvider | undefined
 let gethWallet: RpcWallet | undefined
 let contractKit: ContractKit | undefined
 
-export function openIpcProvider() {
-  if (!ipcProvider) {
-    Logger.debug(`${tag}@openIpcProvider`, `Initializing IPC connection`)
-    ipcProvider = getIpcProvider()
-  }
-  return ipcProvider
-}
+export function* initContractKit() {
+  let retries = 3
+  // Wrap init in retries to handle cases where Geth is initialized but the
+  // IPC is not ready yet. Without changing Geth + RN-Geth, we have no way to
+  // listen for this readiness
+  while (retries > 0) {
+    try {
+      // The kit must wait for Geth to be initialized because
+      // Geth is required for the RpcWallet
+      yield call(waitForGethInitialized)
 
-export function openWallet() {
-  if (!gethWallet) {
-    Logger.debug(`${tag}@openWallet`, `Initializing wallet`)
-    gethWallet = new RpcWallet(openIpcProvider())
-  }
-  return gethWallet
-}
+      if (contractKit || ipcProvider || gethWallet) {
+        throw new Error('Kit not properly destroyed')
+      }
 
-export function openContractKit(fornoMode: boolean) {
-  if (!contractKit) {
-    Logger.info(
-      `${tag}@openContractKit`,
-      `Initializing contractkit, platform: ${Platform.OS}, forno mode: ${fornoMode}`
-    )
-    const web3 = new Web3(fornoMode ? getHttpProvider(DEFAULT_FORNO_URL) : openIpcProvider())
-    const wallet = openWallet()
-    contractKit = newKitFromWeb3(web3, wallet)
+      const fornoMode = yield select(fornoSelector)
+      Logger.info(`${TAG}@initContractKit`, `Initializing contractkit, forno mode: ${fornoMode}`)
+
+      ipcProvider = getIpcProvider()
+      const web3 = new Web3(fornoMode ? getHttpProvider(DEFAULT_FORNO_URL) : ipcProvider)
+
+      Logger.info(`${TAG}@initContractKit`, 'Initializing wallet')
+      gethWallet = new RpcWallet(ipcProvider)
+      yield call([gethWallet, gethWallet.init])
```

Comment:
it seems like we lose the progress of previous retries with this pattern

Reply:
Correct. Seemed like the safer thing to do. Do you think it'll leave behind state that'll be problematic?